objects/src/notes/metadata.rs

```diff
@@ -132,11 +134,77 @@ impl Serializable for NoteMetadata {
 impl Deserializable for NoteMetadata {
     fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
         let sender = AccountId::read_from(source)?;
-        let note_type = NoteType::read_from(source)?;
+        let (note_type, note_execution_hint) = unmerge_type_and_hint(source.read_u64()?)
+            .map_err(|err| DeserializationError::InvalidValue(err.to_string()))?;
         let tag = NoteTag::read_from(source)?;
         let aux = Felt::read_from(source)?;
 
-        Self::new(sender, note_type, tag, aux)
+        Self::new(sender, note_type, tag, note_execution_hint, aux)
             .map_err(|err| DeserializationError::InvalidValue(err.to_string()))
     }
 }
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// Encodes `note_type` and `note_execution_hint` into a [u64] such that the resulting number has
+/// the following structure (from most significant bit to the least significant bit):
+///
+/// - Bits 39 to 36 (4 bits): NoteType
+/// - Bits 35 to 32 (4 bits): NoteExecutionHint tag
+/// - Bits 31 to 0 (32 bits): NoteExecutionHint payload
```

Comment:
Question: why allocate 4 bits to `NoteType`? We only need 2 bits for it, right? Or basically, this could have been something like this:
```rust
/// - Bits 39 to 36 (2 bits): NoteType
/// - Bits 35 to 32 (6 bits): NoteExecutionHint tag
/// - Bits 31 to 0 (32 bits): NoteExecutionHint payload
```

Reply:
I was thinking it would be 4 bits (to account for potential future types) due to [this comment](https://github.com/0xPolygonMiden/miden-base/issues/582#issuecomment-2261570489) but now that I'm re-reading it in this context, maybe you were referring to the note execution tag? Although in your above comment you are suggesting 6 bits for this field so maybe it wasn't that either.

Reply:
Hmmm - maybe I was thinking about future potential types as well - though, it could have also been just a typo :)

I think I'm OK for using just 2 bits for it. If we need to add more types in the future, we'd probably give meaning to type `00` to indicate that additional type information is in some other bits in the metadata (so, type would essentially have variable-length encoding).

Reply:
Got it. Just to be clear, you still think we should save 6 bits for the note execution tag, right? So it would still be 40 bits total.

Reply:
We can also make it 4 - but 6 should be OK too. I'm expecting we'll reshuffle things a bit once we change the size of account ID - but we'll deal with that later.